File: src/main/java/forpleuvoir/hiirosakura/client/feature/event/base/Event.java
package forpleuvoir.hiirosakura.client.feature.event.base;

import com.google.common.collect.ImmutableMap;
import forpleuvoir.hiirosakura.client.HiiroSakuraClient;

import java.util.function.Consumer;

/**
 * 事件
 *
 * @author forpleuvoir
 * <p>#project_name hiirosakura
 * <p>#package forpleuvoir.hiirosakura.client.feature.event.base
 * <p>#class_name Event
 * <p>#create_time 2021-07-23 13:25
 */
public abstract class Event {
    public final HiiroSakuraClient hs = HiiroSakuraClient.getInstance();

    @SuppressWarnings({"rawtypes", "unchecked"})
    public void broadcastHandle(
            ImmutableMap<Class<? extends Event>, ImmutableMap<String, Consumer<? extends Event>>> eventListeners
    ) {
        if (eventListeners.containsKey(this.getClass())) {
            for (Consumer value : eventListeners.get(this.getClass()).values()) {
                value.accept(this);
            }
        }
    }
}

File: src/main/kotlin/forpleuvoir/hiirosakura/client/feature/event/base/Event.kt
package forpleuvoir.hiirosakura.client.feature.event.base

import com.google.common.collect.ImmutableMap
import forpleuvoir.hiirosakura.client.HiiroSakuraClient

/**
 * 事件
 *
 * @author forpleuvoir
 *
 * #project_name hiirosakura
 *
 * #package forpleuvoir.hiirosakura.client.feature.event.base
 *
 * #class_name Event
 *
 * #create_time 2021-07-23 13:25
 */
abstract class Event {
	val hs = HiiroSakuraClient
	fun broadcastHandle(
		eventListeners: ImmutableMap<Class<out Event>, ImmutableMap<String, (Event) -> Unit>>
	) {
		if (eventListeners.containsKey(this.javaClass)) {
			for (value in eventListeners[this.javaClass]!!.values) {
				value.invoke(this)
			}
		}
	}
}